### PR TITLE
Muon efficiency measurement for GEM

### PR DESCRIPTION
**Pull request by qyguo**
Hi all experts, it is the code for Muon efficiency measurement for GEM detector. It is inherited from TnP framework of L. Lunerti (INFN Bologna) following the other sub-detector.
I did the test based on the CMSSW_13_1_0_pre1 framework.

We would like to add this in the next release for P5 and may I ask whether we need to make a backport for this?

### CODE REVIEW

**Comment by watson-ij** [comment]
please test

**Comment by emanueleusai** [comment]
type gem

**Comment by emanueleusai** [comment]
hold

- waiting for 13_0 backport and P5 tests

**Comment by emanueleusai** [comment]
unhold

- PR is not for online dqm

**Comment by emanueleusai** [comment]
+1

**Comment by emanueleusai** [comment]
(differences in EGammaV unrelated to this PR)

File: DQMOffline/MuonDPG/plugins/GEMTnPEfficiencyTask.cc [hunk 1]
```diff
@@ -0,0 +1,848 @@
...
+    std::vector<int> probe_ME0_sta;
+    std::vector<int> probe_ME0_lay;
+    std::vector<int> probe_ME0_chamber;
+    std::vector<float> probe_ME0_pt;
+    std::vector<float> probe_ME0_eta;
+    std::vector<float> probe_ME0_phi;
+    std::vector<float> probe_ME0_dx;
+    uint8_t ME0_stationMatching = 0;
+
+    float gem_matched = false;  // fill detailed plots only for probes matching GEM
```
Comment (perrotta) [inline DQMOffline/MuonDPG/plugins/GEMTnPEfficiencyTask.cc:528]:
```suggestion
    bool gem_matched = false;  // fill detailed plots only for probes matching GEM
```

File: DQMOffline/MuonDPG/plugins/GEMTnPEfficiencyTask.cc [hunk 2]
```diff
@@ -0,0 +1,848 @@
...
+
+          int roll = chId.roll();
+          int region = chId.region();
+          int station = chId.station();
+          int layer = chId.layer();
+          int chamber = chId.chamber();
+          float pt = (*muons).at(i).pt();
+          float eta = (*muons).at(i).eta();
+          float phi = (*muons).at(i).phi();
+
```
Comment (perrotta) [inline DQMOffline/MuonDPG/plugins/GEMTnPEfficiencyTask.cc:546]: I think they can all be made `const` (but please check)
```suggestion
          const int roll = chId.roll();
          const int region = chId.region();
          const int station = chId.station();
          const int layer = chId.layer();
          const int chamber = chId.chamber();
          const float pt = (*muons).at(i).pt();
          const float eta = (*muons).at(i).eta();
          const float phi = (*muons).at(i).phi();

```

File: DQMOffline/MuonDPG/plugins/GEMTnPEfficiencyTask.cc [hunk 3]
```diff
@@ -0,0 +1,848 @@
...
+          m_histos.find("GEM_nPassingProbe_chamber_1D")->second->Fill(GEM_chamber);
+          if (GEM_region < 0) {
+            if (GEM_sta == 2 and GEM_lay == 2)
+              m_histos.find("GEM_nPassingProbe_Ch_region_layer_phase2")->second->Fill(0, GEM_chamber);
+            if (GEM_sta == 2 and GEM_lay == 1)
+              m_histos.find("GEM_nPassingProbe_Ch_region_layer_phase2")->second->Fill(1, GEM_chamber);
+            if (GEM_sta == 1 and GEM_lay == 2)
+              m_histos.find("GEM_nPassingProbe_Ch_region_layer_phase2")->second->Fill(2, GEM_chamber);
+            if (GEM_sta == 1 and GEM_lay == 1)
+              m_histos.find("GEM_nPassingProbe_Ch_region_layer_phase2")->second->Fill(3, GEM_chamber);
```
Comment (perrotta) [inline DQMOffline/MuonDPG/plugins/GEMTnPEfficiencyTask.cc:732]:
```suggestion
            else if (GEM_sta == 2 and GEM_lay == 1)
              m_histos.find("GEM_nPassingProbe_Ch_region_layer_phase2")->second->Fill(1, GEM_chamber);
            else if (GEM_sta == 1 and GEM_lay == 2)
              m_histos.find("GEM_nPassingProbe_Ch_region_layer_phase2")->second->Fill(2, GEM_chamber);
            else if (GEM_sta == 1 and GEM_lay == 1)
              m_histos.find("GEM_nPassingProbe_Ch_region_layer_phase2")->second->Fill(3, GEM_chamber);
```

File: DQMOffline/MuonDPG/plugins/GEMTnPEfficiencyTask.cc [hunk 4]
```diff
@@ -0,0 +1,848 @@
...
+          }
+          if (GEM_region > 0) {
+            if (GEM_sta == 1 and GEM_lay == 1)
+              m_histos.find("GEM_nPassingProbe_Ch_region_layer_phase2")->second->Fill(6, GEM_chamber);
+            if (GEM_sta == 1 and GEM_lay == 2)
+              m_histos.find("GEM_nPassingProbe_Ch_region_layer_phase2")->second->Fill(7, GEM_chamber);
+            if (GEM_sta == 2 and GEM_lay == 1)
+              m_histos.find("GEM_nPassingProbe_Ch_region_layer_phase2")->second->Fill(8, GEM_chamber);
+            if (GEM_sta == 2 and GEM_lay == 2)
+              m_histos.find("GEM_nPassingProbe_Ch_region_layer_phase2")->second->Fill(9, GEM_chamber);
```
Comment (perrotta) [inline DQMOffline/MuonDPG/plugins/GEMTnPEfficiencyTask.cc:742]:
```suggestion
            else if (GEM_sta == 1 and GEM_lay == 2)
              m_histos.find("GEM_nPassingProbe_Ch_region_layer_phase2")->second->Fill(7, GEM_chamber);
            else if (GEM_sta == 2 and GEM_lay == 1)
              m_histos.find("GEM_nPassingProbe_Ch_region_layer_phase2")->second->Fill(8, GEM_chamber);
            else if (GEM_sta == 2 and GEM_lay == 2)
              m_histos.find("GEM_nPassingProbe_Ch_region_layer_phase2")->second->Fill(9, GEM_chamber);
```

File: DQMOffline/MuonDPG/plugins/GEMTnPEfficiencyTask.cc [hunk 5]
```diff
@@ -0,0 +1,848 @@
...
+          m_histos.find("GEM_nFailingProbe_chamber_1D")->second->Fill(GEM_chamber);
+          if (GEM_region < 0) {
+            if (GEM_sta == 2 and GEM_lay == 2)
+              m_histos.find("GEM_nFailingProbe_Ch_region_layer_phase2")->second->Fill(0, GEM_chamber);
+            if (GEM_sta == 2 and GEM_lay == 1)
+              m_histos.find("GEM_nFailingProbe_Ch_region_layer_phase2")->second->Fill(1, GEM_chamber);
+            if (GEM_sta == 1 and GEM_lay == 2)
+              m_histos.find("GEM_nFailingProbe_Ch_region_layer_phase2")->second->Fill(2, GEM_chamber);
+            if (GEM_sta == 1 and GEM_lay == 1)
+              m_histos.find("GEM_nFailingProbe_Ch_region_layer_phase2")->second->Fill(3, GEM_chamber);
```
Comment (perrotta) [inline DQMOffline/MuonDPG/plugins/GEMTnPEfficiencyTask.cc:794]:
```suggestion
            else if (GEM_sta == 2 and GEM_lay == 1)
              m_histos.find("GEM_nFailingProbe_Ch_region_layer_phase2")->second->Fill(1, GEM_chamber);
            else if (GEM_sta == 1 and GEM_lay == 2)
              m_histos.find("GEM_nFailingProbe_Ch_region_layer_phase2")->second->Fill(2, GEM_chamber);
            else if (GEM_sta == 1 and GEM_lay == 1)
              m_histos.find("GEM_nFailingProbe_Ch_region_layer_phase2")->second->Fill(3, GEM_chamber);
```

File: DQMOffline/MuonDPG/plugins/GEMTnPEfficiencyTask.cc [hunk 6]
```diff
@@ -0,0 +1,848 @@
...
+          }
+          if (GEM_region > 0) {
+            if (GEM_sta == 1 and GEM_lay == 1)
+              m_histos.find("GEM_nFailingProbe_Ch_region_layer_phase2")->second->Fill(6, GEM_chamber);
+            if (GEM_sta == 1 and GEM_lay == 2)
+              m_histos.find("GEM_nFailingProbe_Ch_region_layer_phase2")->second->Fill(7, GEM_chamber);
+            if (GEM_sta == 2 and GEM_lay == 1)
+              m_histos.find("GEM_nFailingProbe_Ch_region_layer_phase2")->second->Fill(8, GEM_chamber);
+            if (GEM_sta == 2 and GEM_lay == 2)
+              m_histos.find("GEM_nFailingProbe_Ch_region_layer_phase2")->second->Fill(9, GEM_chamber);
```
Comment (perrotta) [inline DQMOffline/MuonDPG/plugins/GEMTnPEfficiencyTask.cc:804]:
```suggestion
            else if (GEM_sta == 1 and GEM_lay == 2)
              m_histos.find("GEM_nFailingProbe_Ch_region_layer_phase2")->second->Fill(7, GEM_chamber);
            else if (GEM_sta == 2 and GEM_lay == 1)
              m_histos.find("GEM_nFailingProbe_Ch_region_layer_phase2")->second->Fill(8, GEM_chamber);
            else if (GEM_sta == 2 and GEM_lay == 2)
              m_histos.find("GEM_nFailingProbe_Ch_region_layer_phase2")->second->Fill(9, GEM_chamber);
```

File: DQMOffline/MuonDPG/python/gemTnPEfficiencyTask_cfi.py
Comment (perrotta) [inline DQMOffline/MuonDPG/python/gemTnPEfficiencyTask_cfi.py:1]: At some point a `fillDescriptions` method should be added to `BaseTnPEfficiencyTask`, maybe not in this PR though....

File: DQMOffline/MuonDPG/plugins/GEMTnPEfficiencyTask.cc [hunk 7]
```diff
@@ -0,0 +1,848 @@
...
+ * \interited from the TnP framework of  
+ * \author L. Lunerti - INFN Bologna
+ *
+ */
+
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "DataFormats/MuonReco/interface/MuonSegmentMatch.h"
+#include "DataFormats/MuonReco/interface/MuonGEMHitMatch.h"
+#include "DataFormats/Math/interface/deltaPhi.h"
```
Comment (perrotta) [inline DQMOffline/MuonDPG/plugins/GEMTnPEfficiencyTask.cc:14]: Is this used?

Comment (qyguo) [reply DQMOffline/MuonDPG/plugins/GEMTnPEfficiencyTask.cc]: Hi Perrotta, the deltaPhi is not used and I removed it. MuonGEMHitMatch is used to calculate the dx between the propagated track and the hit. and the MuonSegmentMatch is used for the calculation of the propagated track and the segment. So I keep the two.

File: DQMOffline/MuonDPG/plugins/GEMTnPEfficiencyTask.cc (hunk 6 above)
Comment (qyguo) [reply DQMOffline/MuonDPG/plugins/GEMTnPEfficiencyTask.cc]: Thanks and I have changed and pushed.

**Comment by perrotta** [comment]
please test

**Comment by watson-ij** [comment]
I'm not sure if this PR could cause the RelVal-INPUT to timeout?

**Comment by emanueleusai** [comment]
please test

**Comment by perrotta** [comment]
> I'm not sure if this PR could cause the RelVal-INPUT to timeout?

No it doesn't: those timeouts are unfortunately appearing rather frequently in tests, and unless the PR touches exactly those workflows there is normally no need to waste resources and run the tests once again

**Comment by emanueleusai** [comment]
+1

**Comment by perrotta** [comment]
+1

**Comment by perrotta** [comment]
merge

**Pull request merged**